protocols/autonatv2/src/client/handler/request.rs

```diff
@@ -0,0 +1,302 @@
+    #[error("timeout")]
+    Timeout(#[from] futures_bounded::Timeout),
+    #[error("server rejected dial request")]
+    ServerRejectedDialRequest,
+    #[error("server chose not to dial any provided address")]
+    ServerChoseNotToDialAnyAddress,
+    #[error("server ran into an internal error")]
+    InternalServerError,
+    #[error("server did not respond correctly to dial request")]
+    InvalidResponse,
+    #[error("server was unable to connect to address: {addr:?}")]
+    UnableToConnectOnSelectedAddress { addr: Option<Multiaddr> },
+    #[error("server experienced failure during dial back on address: {addr:?}")]
+    FailureDuringDialBack { addr: Option<Multiaddr> },
+}
+
+
+#[derive(Debug)]
+pub struct TestEnd {
+    pub dial_request: DialRequest,
+    pub suspicious_addr: Vec<Multiaddr>,
+    pub reachable_addr: Multiaddr,
+}
+
+#[derive(Debug)]
+pub enum ToBehaviour {
+    TestCompleted(Result<TestEnd, Error>),
+    PeerHasServerSupport,
+}
+
+#[derive(Debug)]
+pub enum FromBehaviour {
+    PerformRequest(DialRequest),
+}
+
+pub struct Handler {
+    queued_events: VecDeque<
+        ConnectionHandlerEvent<
+            <Self as ConnectionHandler>::OutboundProtocol,
+            <Self as ConnectionHandler>::OutboundOpenInfo,
+            <Self as ConnectionHandler>::ToBehaviour,
+        >,
+    >,
+    outbound: futures_bounded::FuturesSet<Result<TestEnd, Error>>,
+    queued_requests: VecDeque<DialRequest>,
+}
+
+impl Handler {
+    pub fn new() -> Self {
+        Self {
+            queued_events: VecDeque::new(),
+            outbound: FuturesSet::new(DEFAULT_TIMEOUT, DEFAULT_MAXIMUM_CAPACITY),
+            queued_requests: VecDeque::new(),
+        }
+    }
+}
+
+impl ConnectionHandler for Handler {
+    type FromBehaviour = FromBehaviour;
+
+    type ToBehaviour = ToBehaviour;
+
+    type InboundProtocol = DeniedUpgrade;
+
+    type OutboundProtocol = ReadyUpgrade<StreamProtocol>;
+
+    type InboundOpenInfo = ();
+
+    type OutboundOpenInfo = ();
+
+    fn listen_protocol(&self) -> SubstreamProtocol<Self::InboundProtocol, Self::InboundOpenInfo> {
+        SubstreamProtocol::new(DeniedUpgrade, ())
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<
+        ConnectionHandlerEvent<Self::OutboundProtocol, Self::OutboundOpenInfo, Self::ToBehaviour>,
+    > {
+        if let Some(event) = self.queued_events.pop_front() {
+            return Poll::Ready(event);
+        }
+        if let Poll::Ready(m) = self.outbound.poll_unpin(cx) {
+            return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(
+                ToBehaviour::TestCompleted(m.map_err(Error::Timeout).and_then(identity)),
+            ));
+        }
+        if !self.queued_requests.is_empty() {
+            return Poll::Ready(ConnectionHandlerEvent::OutboundSubstreamRequest {
+                protocol: SubstreamProtocol::new(REQUEST_UPGRADE, ()),
+            });
+        }
```

Comment:
This is going to open way more streams than you think because you are not tracking the state, which requests you've already requested a stream for. You need to split `queued_requests` into two lists.

Reply:
I've been exploring patterns on how to handle this better and I think my favourite so far is to use a `oneshot`. See https://github.com/libp2p/rust-libp2p/pull/4834/files#diff-8b77658182acf6d24228f0a9ff05e0fb5b8b9a7be031029130969f4772000694.

Reply:
I think that is resolved too.